### ISSUE TITLE
Implementation of SWP-08 routing protocol

### ISSUE THREAD

**Issue by evenover** [bitfocus/companion-module-requests#502]
Hey!

I would like companion to implement the SWP-08 router protocol. Its a protocol supported by alot of different router-systems like ross ultrix, direct out matrix-routing and lots more.
I want to be able to perform routes and read out routing info. Then be able to use the feedback from the router to make a tallysystem.

Documentation apllies to both SWP-02 and 08
https://wwwapps.grassvalley.com/docs/Manuals/sam/Protocols%20and%20MIBs/Router%20Control%20Protocols%20SW-P-88%20Issue%204b.pdf

**Comment by daniep01** [comment]
I will make a start on this module very soon

**Comment by evenover** [comment]
Awesome! Keep me updated and i can run tests

**Comment by daniep01** [comment]
Thanks! Are you able to test on the Ultrix or GV frames?

**Comment by evenover** [comment]
Yes, i can test with ultrix.

**Comment by pomerle** [comment]
This would be a great addition. 

You can also use rosstalk on an Ultrix version 4.5 and up for crosspoint control I have it working however the Companion UI appears to send an extra 3a (:)  command.   
I found that I also needed to Identify the panel ID  (I) this is what I did to get source 2 routed to Destination 2. 
  
![image](https://user-images.githubusercontent.com/86025288/122296332-487f1080-cec8-11eb-9920-59ab62e00869.png)

**Comment by daniep01** [comment]
Hi,

I have made a test build which implements the make crosspoint command. I think it should be obvious how to use it, any feedback or comments welcome. I am planning to implement the get crosspoint command next.

[https://we.tl/t-BQNSZvHJuT](https://we.tl/t-BQNSZvHJuT)

**Comment by evenover** [comment]
I've now tested it against the ultrix and it works very well. awesome job! 👍 

I want to be able to have one button to choose destination and level and then a button with a the source I want to route. Like how the bmd videohub module is working also with feedback. this way i can be able to route different sources to each level on my ultrix.
It will be more flexible to build pages with destinations and sources.

Keep me updated and i can keep testing and give feedback.

**Comment by daniep01** [comment]
Yes, make sense. Here's a build with the extra commands for presetting level and destination.

[https://we.tl/t-VUc2HWxSuL](https://we.tl/t-VUc2HWxSuL)

**Comment by evenover** [comment]
Nice, looks really good :) I'm not by my ultrix at the moment, but will the "get cross-point" bring back the name from the router also? the protocoll supports up to 12 character names i think.

Is it possible to have a statusbutton in addition to the "get crosspoint" that gets the info from the current selected destination?

**Comment by daniep01** [comment]
It would be great to get names but I can't see how to do it that in the spec you posted. Can you point me to where it's documented?

**Comment by evenover** [comment]
I could not find it in the document i provided, but does this help?
https://www.directout.eu/upload/m1k2/www_ship/en/help/swp08.html

I can try to look more on the internet for something more if this doesnt help :)

**Comment by daniep01** [comment]
It's helpful that it confirms it's possible, but it's lacking the detail of the bits and bytes.

**Comment by evenover** [comment]
I understand 👍 Let me see if i can get hold of some more documentation

**Comment by evenover** [comment]
I found a test-client for swp-08 from studer. maybe thats any help? Please let me know

[Probel TestClient SWP08.zip](https://github.com/bitfocus/companion-module-requests/files/6679128/Probel.TestClient.SWP08.zip)

**Comment by daniep01** [comment]
The source code for the test client would be perfect, but I'm guessing it's not available.

**Comment by evenover** [comment]
I'll keep looking!

**Comment by evenover** [comment]
Would it be helpful if I used wireshark to analyze the responses coming from the ultrix?

**Comment by daniep01** [comment]
Do you think the test client can connect to the Ultrix? It only appears to have serial com ports for connection options and the Ultrix expects SWP08 over tcp.

**Comment by evenover** [comment]
I might be able to use a serial to ip-converter.

Or maybe I can try sending what we might think the command for

**Comment by daniep01** [comment]
Couple of updates: 
- I have added decoding of responses from the router when routes are made either in companion or any other control system.
- There are new actions for getting the names. I haven't found any spec for the data that is returned so the parsing is based on guesswork and a few hacks so could well break. The labels also appear to be only 8 chars long which limits the usefulness a bit.

Both of these just write information to the companion debug log for now, please have a look and let me know if these are working for you.

https://we.tl/t-nI8fBVx0UZ

**Comment by chibiconsulting** [comment]
I wonder if the extended commands for larger matrix sizes might also return the longer character labels? In any case supporting larger matrix sizes would be useful.

Notes from the page: https://www.directout.eu/upload/m1k2/www_ship/en/help/swp08.html
228[0xE4]	Extended All Source Names Request	
229[0xE5]	Extended Single Source Name Request	
230[0xE6]	Extended All Destinations Association Names Request	
231[0xE7]	Extended Single Destination Association Names Request	

Also as a side note; support for UMD labels could be useful; since, if defined in the router, those UMD labels should already be optimized for use inside buttons or small displays. * Not all routers support that field so it may not work everywhere.
104[0x68]	All UMD Labels Request	
105[0x69]	Single UMD Label Request
232[0xE8]	Extended All UMD Labels Request	
233[0xE9]	Extended Single UMD Label Request

**Comment by evenover** [comment]
I will test out tomorrow and give feedback on how it looks 🙌

I will try to find full documentation through grass valley 😊

**Comment by evenover** [comment]
Now i have been testing and the logs is saying that your guesswork is right 👍 

My ultrix-setup got alot of sources(500) and destinations(200) so it takes some time for it to send all the labels, but they are coming in named as in the utrix. Some names are over 8char so it cuts them, but thats okay, i can rename them. We could also try the extended labels as mentioned by @chibiconsulting. It's described in the ultrix-manual that it is supported.

As a feedback i think we should define the size of the matrix in the "edit module"  this way the module can scale with variables to contain the labels of the sources and destinations. We could probably have the load labels in the edit-tab as well? because i do not think this is something that will change much on the fly and you should not need it on the surface itself.

It should also be variables that store the xpoints with label of the source.

Will there be feedback on which destination is selected? When i press a destinationbutton it should be able to light up, then i will know which destination i'm routing to.

**Comment by daniep01** [comment]
Thanks for the confirmation. It's useful to know the size you are working with as currently I have limited it to 128x128 and 16 levels. To do more levels or more than 1024x1024 requires the extended commands. Hopefully GV can provide some more information on the extended commands. 

I think a lot of this can become more automatic, by asking for the labels when the module connects it will know the size of the router and the names can make a drop down list to pick from. I will probably remove the get names actions from the user interface.

Feedbacks to change button colours is definitely on the todo list. I'm also wondering if we need an easy way of routing multiple levels - do you have separate levels for SDI and audio which need to be routed simultaneously?

**Comment by evenover** [comment]
It would be cool to have a "married"-command where i can route all levels on a source. My router i set up with 17 layers because: SDI video +16 audio tracks. but for starters; 8 level control will work just fine. I'm also trying to get in touch with ross and see if they can provide info on the protocol.

If the module asks for labels at connection i think it would be nice if it had a status ON if labels are completely loaded. The log actually loaded all of the labels up to 500

**Comment by daniep01** [comment]
Coloured backgrounds for the selected level, destination, and source can now be done by adding feedbacks to the buttons.

**Comment by daniep01** [comment]
> Regarding levels, it looks like there are only 4 bits (Bits[0-3] of Byte 1) allocated to level in SW-P-08 so 16 levels would probably the most for this protocol.

Yes, absolutely true, I think that's why the extended protocol commands are needed.

**Comment by evenover** [comment]
I think I will get hold of the documentation in the next few days. I'll keep you updated.

Regarding the levels I think 9 levels would be enough in most cases. Video + 8 audio. We have to see if there is anything about it in the extended documentation.

**Comment by daniep01** [comment]
New  build:

[https://we.tl/t-o4fsRCp4nV](https://we.tl/t-o4fsRCp4nV)

- Getting names should be about 5 times quicker
- Sources and Destinations up to 1024 now supported by implementing the multiplier bits in byte 2 of the standard set/get messages (please test high numbers)
- Feedback to change background colour for selected source, destination, level
- Variables for selected source, destination, level
- Variables for name counts (ignore this, it's not working)

**Comment by chibiconsulting** [comment]
For 2.2.0 (2.2.0-061e92b1-3423) Tested in companion emulator only, My protocol emulator is not making connections at the moment.

Names not tested. BTW, when my testing setup works, where will the names show up?
For the destination and sources including high numbers; the feedback works as expected.
The implemented variables seem to work as expected.

General note/question.
For destination, level, and source, the allowed values start at 1.
All of these start at 0 in the protocol.
Are these shifted in code or should they allow values from 0? Example: levels 0-15 vs levels 1-16.

Levels question.
Operationally, the levels are usually multiselect with the ability to toggle each level on or off.
Will this eventually be possible with this implementation?

**Comment by evenover** [comment]
I'm on vacation at the moment, but if i recieve the documentation I will post it here. Will be back at testing Monday 28th.

**Comment by daniep01** [comment]
> Names not tested. BTW, when my testing setup works, where will the names show up?

In the current build they only appear in the debug log. I plan to use them for dropdown lists to select sources and destinations in the next build.

> General note/question.
> For destination, level, and source, the allowed values start at 1.
> All of these start at 0 in the protocol.
> Are these shifted in code or should they allow values from 0? Example: levels 0-15 vs levels 1-16.

I think users will expect them to start at 1 so I add/subtract in the code.

> Levels question.
> Operationally, the levels are usually multiselect with the ability to toggle each level on or off.
> Will this eventually be possible with this implementation?

It would make sense but I need to figure out a way of doing this in the companion user interface. I don't think 17 or even 9 checkboxes in an action is viable.

**Comment by daniep01** [comment]
@chibiconsulting Does your emulator reply to command 0x61 (implementation request)? My module will ask this as soon as it connects and expect a list of implemented commands to be returned.

**Comment by chibiconsulting** [comment]
Levels: Normally I would expect level controls to either be a few buttons on the routing page (the vast majority of router installations use just 1 to 4 levels) or a separate page or pages with all of the available level selections for those with complex routing needs.

Emulator: It is an old very basic emulator and does not respond to the 0x61 request at all. I suspect it does not implement some of the commands in use with the current version.

**Comment by daniep01** [comment]
In this build all of the level controls are now multi-select. You will need to go to the module settings page and configure the number of levels you wish to see in this list before it can be used. I have also added select by name for source and destination.

[https://we.tl/t-1d7kDbPCqw](https://we.tl/t-1d7kDbPCqw)

**Comment by daniep01** [comment]
@josephdadams could you assign this to me, and maybe move it to the module repo now we have one?
